UCX osc: add support for acc_single_intrinsic

devreal opened the pull request:
This PR adds support for the `acc_single_intrinsic` info key / mca parameter that allows the user to specify that only a single intrinsic type will be used in accumulate operations to avoid taking the accumulate lock for every operation. It also enables asynchronicity for accumulate operations that do not require emulation through compare-and-swap (currently only `replace` and `sum` and `compare-and-swap` itself).

Limitations:

- ~128 bit types are not supported and lead to an error if this info key is set (not sure how to handle 128 bit values that with the current UCX API)~
- ~Operations other than `replace` and `sum` in `MPI_Accumulate` require a `get` followed by a `cas` loop until successful replacement. This loop is currently executed directly but could be encapsulated in a UCX completion request, issuing a new cas if the previous one failed. Not sure whether UCX allows new operations to be issued from within a completion callback though.~
- ~After merging of #6933 there might be a few more chances to reduce latency, e.g., by fencing the release of the accumulate lock with the completion of the final put (if `acc_single_intrinsic` is not enabled).~

artpol84 commented:
@devreal, sorry for the delay
I'll try to review later this week and will ask @janjust to go over it next week (as h isn't available this week).

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -582,13 +818,29 @@ int ompi_osc_ucx_compare_and_swap(const void *origin_addr, const void *compare_a
+    }
+
+    uint64_t compare_val;
+    memcpy(&compare_val, compare_addr, dt_bytes);
```
Is this to ensure 64bit alignment?

devreal replied:
Yeah, I've seen 64bit accesses of potentially arbitrarily aligned pointers provided by the user in several places (here and elsewhere in Open MPI, iirc). I'm not sure whether that is safe on all architectures (it is on x86 and seems to be on Arm >= v7). The C language defines it as UB if a pointer references a memory location that is not suitably aligned for the referenced type (www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf, page 74)

devreal replied:
There is also this nugget hidden in `opal_common_ucx_atomic_cswap`:

```C
        if (op_size == sizeof(uint64_t)) {
            *(uint64_t*)result = tmp;
        } else {
            assert(op_size == sizeof(uint32_t));
            *(uint32_t*)result = tmp;
        }
```
https://github.com/open-mpi/ompi/blob/master/opal/mca/common/ucx/common_ucx.h#L217

I can only guess that the intention there was to avoid unaligned accesses? I don't see language in the standard that prohibits `MPI_Compare_and_swap` to be called on `MPI_SHORT`.

Since we have only four relevant data type sizes to cover (1, 2, 4, 8) we could have an inlined custom `memcpy` that issues loads on appropriately typed pointers, if there is a concern about using `memcpy`.

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -582,13 +818,29 @@ int ompi_osc_ucx_compare_and_swap(const void *origin_addr, const void *compare_a
+    uint64_t compare_val;
+    memcpy(&compare_val, compare_addr, dt_bytes);
+    memcpy(result_addr, origin_addr, dt_bytes);
+    ret = opal_common_ucx_wpmem_fetch_nb(module->mem, UCP_ATOMIC_FETCH_OP_CSWAP,
```
Could you clarify, why we need to replace `opal_common_ucx_wpmem_cmpswp`?

devreal replied:
There is no `nb` version of `opal_common_ucx_wpmem_cmpswp`, it calls `opal_common_ucx_atomic_cswap` (and in turn `opal_common_ucx_atomic_fetch`) and eventually waits for the operation to complete.

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -323,6 +323,215 @@ static inline int get_dynamic_win_info(uint64_t remote_addr, ompi_osc_ucx_module
+    ompi_datatype_type_size(origin_dt, &origin_dt_bytes);
+    ompi_datatype_type_size(target_dt, &target_dt_bytes);
+
+    if (origin_dt_bytes  > sizeof(uint64_t) ||
```
Should we also have ompi_datatype_is_predefined(origin_dt) check?

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -323,6 +323,215 @@ static inline int get_dynamic_win_info(uint64_t remote_addr, ompi_osc_ucx_module
+            user_req_ptr = ucx_req;
+            // issue a fence if this is the last but not the only element
+            if (0 < i) {
+                ret = opal_common_ucx_wpmem_fence(module->mem);
```
Currently, this fence is a no-op.
@yosefe do we need to have something stronger to ensure the correct ordering?

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -323,6 +323,215 @@ static inline int get_dynamic_win_info(uint64_t remote_addr, ompi_osc_ucx_module
+            memcpy(&value, origin_addr, origin_dt_bytes);
+        }
+        ret = opal_common_ucx_wpmem_fetch_nb(module->mem, opcode, value, target,
+                                             result_addr ? result_addr : &(module->req_result),
```
Minor: I think the following would be a bit fewer computations:

```C
void *output_ptr = &(module->req_result);
if( result_addr ) {
    output_ptr = result_addr;
}
...
ret = opal_common_ucx_wpmem_fetch_nb(module->mem, opcode, value, target,
                                             output_addr, ... );
...
if (result_addr) {
    output_addr += origin_dt_bytes;
}
```

devreal replied:
I will change that, sounds reasonable.

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -323,6 +323,215 @@ static inline int get_dynamic_win_info(uint64_t remote_addr, ompi_osc_ucx_module
+            return ret;
+        }
+
+        ret = opal_common_ucx_wpmem_flush(module->mem, OPAL_COMMON_UCX_SCOPE_EP, target);
```
You also will need to flush it per element.

artpol84 commented:
Overall, this looks good to me. Though I'd like to double-check some details.

devreal commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -323,6 +323,215 @@ static inline int get_dynamic_win_info(uint64_t remote_addr, ompi_osc_ucx_module
+    ompi_datatype_type_size(origin_dt, &origin_dt_bytes);
+    ompi_datatype_type_size(target_dt, &target_dt_bytes);
+
+    if (origin_dt_bytes  > sizeof(uint64_t) ||
```
I'd expect that this is checked upon entry to MPI. Have to check if that happens.

devreal commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -323,6 +323,215 @@ static inline int get_dynamic_win_info(uint64_t remote_addr, ompi_osc_ucx_module
+        uint64_t target_val = 0;
+
+        // get the value from the origin
+        ret = opal_common_ucx_wpmem_putget(module->mem, OPAL_COMMON_UCX_GET,
```
I agree. Considering that the standard restricts the use of accumulate operations to either `same_op` or `same_op_no_op` we can fall back to taking the accumulate lock for operations that are not supported by UCX (like `MPI_PROD`)

devreal commented:
I addressed the reviewer comments and added two optimizations: the accumulate lock won't be acquired if an exclusive lock is present; and the release of the accumulate lock will be overlapped with any previous operations.

artpol84 commented:
Thanks, I’ll go over later this week

devreal commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -235,37 +235,76 @@ static inline int ddt_put_get(ompi_osc_ucx_module_t *module,
+                return OMPI_ERROR;
+            }
+            if (result_value == TARGET_LOCK_UNLOCKED) {
+                return OMPI_SUCCESS;
```
break here

devreal commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -323,6 +362,110 @@ static inline int get_dynamic_win_info(uint64_t remote_addr, ompi_osc_ucx_module
+        op == &ompi_mpi_op_no_op.op) {
+        size_t dt_bytes;
+        ompi_datatype_type_size(origin_dt, &dt_bytes);
+        if (ompi_datatype_is_predefined(origin_dt) &&
```
`assert` that `origin_dt` is a predefined datatype

devreal replied:
Check for `module->acc_single_intrinsic` here

devreal commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -323,6 +362,110 @@ static inline int get_dynamic_win_info(uint64_t remote_addr, ompi_osc_ucx_module
+    ompi_datatype_type_size(origin_dt, &origin_dt_bytes);
+    ompi_datatype_type_size(target_dt, &target_dt_bytes);
+
+    if (sizeof(uint64_t) > origin_dt_bytes  ||
```
Check if UCX supports 8 and 16bit operands

devreal replied:
See also https://github.com/open-mpi/ompi/issues/6777

devreal commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -582,14 +745,23 @@ int ompi_osc_ucx_compare_and_swap(const void *origin_addr, const void *compare_a
-                                     *(uint64_t *)origin_addr, target,
-                                     result_addr, dt_bytes, remote_addr);
-    if (ret != OMPI_SUCCESS) {
+    if (sizeof(uint64_t) < dt_bytes) {
```
Move this check to the beginning of the function

devreal commented:
@artemry-mlnx I finally found the time to work on this again. I hopefully have addressed all the points we found during our review in December. Please take a look if you can spare some time :) I would be happy if we could get this into 5.0.

@yosefe One point that is not clear to me:

There is a specific function in `opal/mca/common/ucx/common_ucx.h` for compare-and-swap (https://github.com/open-mpi/ompi/blob/master/opal/mca/common/ucx/common_ucx.h#L204). I don't really understand why this code is needed (and doubt that it's correct):

```C
static inline
int opal_common_ucx_atomic_cswap(ucp_ep_h ep, uint64_t compare,
                                 uint64_t value, void *result, size_t op_size,
                                 uint64_t remote_addr, ucp_rkey_h rkey,
                                 ucp_worker_h worker)
{
    uint64_t tmp = value;
    int ret;

    ret = opal_common_ucx_atomic_fetch(ep, UCP_ATOMIC_FETCH_OP_CSWAP, compare, &tmp,
                                       op_size, remote_addr, rkey, worker);
    if (OPAL_LIKELY(OPAL_SUCCESS == ret)) {
        /* in case if op_size is constant (like sizeof(type)) then this condition
         * is evaluated in compile time */
        if (op_size == sizeof(uint64_t)) {
            *(uint64_t*)result = tmp;
        } else {
            assert(op_size == sizeof(uint32_t));
            *(uint32_t*)result = tmp;
        }
    }
    return ret;
}
```

AFAICS, the assignment from `tmp` after the operation is faulty in the 32 bit case: with `op_size == 4` the upper 4B of `tmp` will be overwritten with the previous value in `remote_addr` and the lower 4B should be left untouched by `UCP_ATOMIC_FETCH_OP_CSWAP`. However, the assignment `*(uint32_t*)result = tmp;` will downcast the value in `tmp`, cutting off the upper 4B and assigning the lower 4B. Am I missing something here? Do we really need a special function for compare-and-swap or can't we just handle it through `opal_common_ucx_atomic_fetch` with `UCP_ATOMIC_FETCH_OP_CSWAP`? See also the comment here https://github.com/open-mpi/ompi/pull/6980#discussion_r340424408

yosefe commented:
@devreal it works because the data is taken from first 4B of the target address and also placed in first 4B of the result. on LE arch, the lower 4B of tmp are overwritten with the previous value in remote_addr, not the upper 4B.

devreal commented:
> it works because the data is taken from first 4B of the target address and also placed in first 4B of tmp. on LE arch, the lower 4B of tmp are overwritten with the previous value in remote_addr, not the upper 4B.

Ahh, I see. I was not thinking in terms of LE architectures. That would not be portable to BE architectures though, right? Why not copy the `value` into `result` and pass the `result` pointer directly to `opal_common_ucx_atomic_fetch`? That would make it easier to create a nonblocking version of compare-and-swap.

In similar terms, it is not entirely clear to me what the format of the 64bit value passed to `ucp_atomic_fetch_nb` is if `op_size == 4`. Is the 64-bit value cast down to a 32-bit integer? Are either the upper or lower 4B used? I don't see that specified in the UCX standard.

yosefe commented:
@devreal it works on BE as well, in this case the first 4B of tmp will be tye "upper" part. However, it doesn't really matter, we don't to integer case on tmp, we just do pointer cast, which does not care about endianness.

- the input value for UCX atomic is "uint64_t" so it's int-casted to uint32 if needed.
- the output value of the atomic is "void*" so it's a direct pointer to 32-bit result

devreal commented:
I updated this PR to include the following:

1) a non-blocking variant of `opal_common_ucx_wpmem_cmpswp`
1) a safe and portable way to load and store 64-bit integer values to variable-size pointers without relying on `memcpy`
2) support for compare-and-swap for types that are not supported by UCX atomics (from what I understand that is 8, 16, and 128-bit types)

devreal commented:
@awlauria I think this PR is now in a good state again (after a long pause). It's awaiting a review now. Can we add this to the list of tracked issues for 5.0?

awlauria commented:
@devreal I will add it to the list. Thanks! However - it looks like there may be a commit or here missing the signed-off by, can you fix that?

devreal commented:
@awlauria Yes, I'm aware of that. I wasn't sure whether some of the commits should be squashed. I will fix that one unsigned today though.

devreal commented:
Fixed two bugs that slipped through during initial testing and squashed down some commits.

artpol84 commented:
@janjust can you please review

janjust commented:
@devreal @artpol84 , guys this looks ok to me -I don't have any constructive comments. But @devreal out of curiosity, what kind of testing was done on this? Do you have any tests of your own that you sued, either way, can you share?

devreal commented:
@janjust I tested this on a ConnectX-3 cluster with a benchmark I wrote that covers most of the MPI RMA functions and runs with both shared and exclusive locks. The code is available at https://github.com/devreal/mpi-progress and I am happy to contribute that to any existing test suite.

awlauria commented:
@artpol84 you still have a change request that is preventing this from getting merged. Can you give it another look when you get a chance? Thanks!

jladd-mlnx commented:
@artpol84 have you and @devreal reached a consensus? If so, can you please remove the request change label, Artem?

jladd-mlnx commented:
@janjust @artpol84 @devreal what is the status? Seems like it's a critical fix. Can we merge it?

artpol84 commented:
I'd like to review one more time. I'll plan to go over it this week. I am sorry for not doing this before.

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -582,20 +780,50 @@ int ompi_osc_ucx_compare_and_swap(const void *origin_addr, const void *compare_a
-    ret = opal_common_ucx_wpmem_cmpswp(module->mem,*(uint64_t *)compare_addr,
-                                     *(uint64_t *)origin_addr, target,
-                                     result_addr, dt_bytes, remote_addr);
+    if (ATOMIC_SIZE_SUPPORTED(dt_bytes)) {
```
```suggestion
    if (ATOMIC_SIZE_SUPPORTED(remote_addr, dt_bytes)) {
```

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -604,16 +832,22 @@ int ompi_osc_ucx_fetch_and_op(const void *origin_addr, void *result_addr,
+    ompi_datatype_type_size(dt, &dt_bytes);
+
+    /* UCX atomics are only supported on 32 and 64 bit values */
+    if (ATOMIC_SIZE_SUPPORTED(dt_bytes) &&
```
```suggestion
    uint64_t remote_addr = (module->addrs[target]) + target_disp * OSC_UCX_GET_DISP(module, target);
    if (ATOMIC_SIZE_SUPPORTED(remote_addr, dt_bytes) &&
```

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -604,16 +832,22 @@ int ompi_osc_ucx_fetch_and_op(const void *origin_addr, void *result_addr,
+        (op == &ompi_mpi_op_no_op.op || op == &ompi_mpi_op_replace.op ||
+         op == &ompi_mpi_op_sum.op)) {
+        uint64_t value;
         uint64_t remote_addr = (module->addrs[target]) + target_disp * OSC_UCX_GET_DISP(module, target);
```
```suggestion
```

artpol84 commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -763,17 +1009,29 @@ int ompi_osc_ucx_get_accumulate(const void *origin_addr, int origin_count,
             if (ret != OMPI_SUCCESS) {
                 return ret;
             }
-
```
Can we remove this?

devreal replied:
What is 'this'? The whitespace? It's part of a longer section of code that was removed.

artpol84 replied:
ok

devreal commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -544,15 +708,53 @@ int ompi_osc_ucx_accumulate(const void *origin_addr, int origin_count,
+        ompi_request_complete(&ucx_req->super, true);
+    }
+
+    return end_atomicity(module, target, lock_acquired, free_ptr);
```
`free_ptr` is passed to `end_atomicity` because the transfer is overlapped with the release of the atomic lock and only after that the `free_ptr` can actually be free'd.

devreal commented on ompi/mca/osc/ucx/osc_ucx_comm.c:
```diff
@@ -24,6 +26,10 @@
 
+/* macro to check whether UCX supports atomic operation on the size the operands */
+#define ATOMIC_SIZE_SUPPORTED(_size) (sizeof(uint32_t) == _size || \
+                                      sizeof(uint64_t) == _size)
```
I have extended `ATOMIC_SIZE_SUPPORTED` to check for proper target memory alignment

artpol84 commented:
@devreal , Do we have everything in place now?
